fast_neural_style/train.py

```diff
@@ -0,0 +1,127 @@
+from torch.autograd import Variable
+from torch.utils.serialization import load_lua
+
+import net as models
+import numpy as np
+import os
+from datetime import datetime
+import cv2
+
+def download_convert_vgg16_model():
+    if not os.path.exists('vgg16feature.pth'):
+        if not os.path.exists('vgg16.t7'):
+            os.system('wget http://cs.stanford.edu/people/jcjohns/fast-neural-style/models/vgg16.t7')
+        vgglua = load_lua('vgg16.t7')
+        vgg = models.vgg16feature()
+        for src,dst in zip(vgglua.parameters()[0],vgg.parameters()):
+            dst[:] = src[:]
+        torch.save(vgg.state_dict(), 'vgg16feature.pth')
+
+def load_image(path, size=None):
+    img = cv2.imread(path).astype('float32')
+    img = cv2.resize(img,(size,size))
+    img = img.transpose(2, 0, 1)
+    return img
+
+def gram_matrix(y):
+    b, ch, h, w = y.size()
+    features = y.view(b, ch, w*h)
+    features_t = features.transpose(1,2)
+    gram = features.bmm(features_t)/(ch*h*w)
+    return gram
+
+# Training settings
+parser = argparse.ArgumentParser(description='PyTorch Fast Neural Style Example')
+parser.add_argument('--epochs', type=int, default=2, metavar='N',
+                    help='number of epochs to train (default: 2)')
+parser.add_argument('--lr', type=float, default=0.001, metavar='LR',
+                    help='learning rate (default: 0.001)')
+parser.add_argument('--no-cuda', action='store_true', default=False,
+                    help='enables CUDA training')
+parser.add_argument('--dataset', '-d', default='../../train2014', type=str,
+                    help='dataset directory path (according to the paper, use MSCOCO 80k images)')
+parser.add_argument('--style_image', '-s', type=str, default='style_2.png',
+                    help='style image path')
+parser.add_argument('--image_size', default=512, type=int)
+parser.add_argument('--lambda_feat', default=1.0, type=float)
+parser.add_argument('--lambda_style', default=5.0, type=float)
+args = parser.parse_args()
+args.cuda = not args.no_cuda and torch.cuda.is_available()
+
+lambda_f = args.lambda_feat
+lambda_s = args.lambda_style
+image_size = args.image_size
+n_epoch = args.epochs
+
+fs = os.listdir(args.dataset)
+fs.sort()
+imagepaths = []
+for fn in fs:
+    base, ext = os.path.splitext(fn)
+    if ext == '.jpg' or ext == '.png':
+        imagepath = os.path.join(args.dataset,fn)
+        imagepaths.append(imagepath)
+n_iter = len(imagepaths)
+
+
+download_convert_vgg16_model()
+model = models.FastStyleNet()
+vgg = models.VGGFeature()
+vgg.load_state_dict(torch.load('vgg16feature.pth'))
+optimizer = optim.Adam(model.parameters(), lr=args.lr)
+if args.cuda:
+    model.cuda()
+    vgg.cuda()
+
+
+style = load_image(args.style_image,image_size) - 120
+style = style.reshape((1,)+style.shape)
+style = torch.from_numpy(style)
+if(args.cuda):
+    style = style.cuda()
+feature_s = vgg(Variable(style, volatile=True))
+gram_s = [gram_matrix(y) for y in feature_s]
```

Comment:
Thanks for your advice.

1. I have used ```cv2.imread``` to load image and opencv's image defaults to BGR.
2. The range of opencv's image defaults to [0,255].
3. I have used [120,120,120] to replace the mean [103.939, 116.779, 123.68]. The preprocessing can be reduced to just one line ```x = x - 120```, the deprocessing can be reduced to ```x = x + 120```. It may be more easy for some starters to read the whole code. After this simplification, there is no any visible difference in the output image.